hazelcast/src/main/java/com/hazelcast/spi/discovery/AbstractDiscoveryStrategy.java

```diff
@@ -52,7 +54,7 @@ public void start() {
     }
 
     @Override
-    public PartitionGroupStrategy getPartitionGroupStrategy() {
+    public PartitionGroupStrategy getPartitionGroupStrategy(Collection<? extends Member> allMembers) {
```

Comment:
I think we should keep the deprecated method here. Because `getPartitionGroupStrategy` now returns empty list instead of null. Otherwise we will create a discrepancy.

Reply:
Unfortunately, that will break the logic where deprecated method works "as is" with new method working as expected.

Reply:
>The intent here is that the original, deprecated method works the same as before.

But it is not working the same as before, now we return an empty-list from parameter-less `getPartitionGroupStrategy` instead of null. For a patch release scope, we should keep old method as is and should leave the decision to override new method to user. Otherwise this will be a breaking change.

Reply:
I am not sure I understand you. Yes, if you take this method in isolation, it is a breaking change.
However, this change is not isolated, and is balanced out by the default methods in the interface and changes in the logic in `SPIAwareMemberGroupFactory.java`.
All tests also pass with deprecated and new methods.

If you leave the abstract method "as it" it will not be possible *not* to override any methods at all, as is possible currently, thus breaking this logic, and the tests will fail.

Perhaps you can suggest an edit that will pass the tests and allow either the deprecated and the "new" method to be overridden with the new method being the default?

Reply:
I mean this test passes with current 4.2.z but not with your PR:

```
@Test
    public void testNoBreakingChangeInApi() {
        CustomDiscoveryStrategy strategy = new CustomDiscoveryStrategy();
        assertNull(strategy.getPartitionGroupStrategy());
    }

    static class CustomDiscoveryStrategy extends AbstractDiscoveryStrategy {

        CustomDiscoveryStrategy() {
            super(null, Collections.<String, Comparable>emptyMap());
        }

        @Override
        public Iterable<DiscoveryNode> discoverNodes() {
            return null;
        }
    }
```

This is a breaking change and we don't do such changes in patch releases.

Reply:
Alright... let's see here :)

Reply:
Sorry, @ahmetmircik, I still don't see an issue here. You are testing the abstract class in isolation (as mentioned above)
This API is used by Hazelcast internals to correctly interpret partition groups.
Since the internals implementation changed to match the new API, I don't see the problem here.
In other words, the actual behavior of the code in totality did not change.

Do you have a way for me to satisfy your desire / requirement and not break the logic where existing code works, new code works also, as far as getting the correct implementation of partition group subclass?
I can't really see another way to implement this...

Reply:
i.e. Default implementation in Abstract class not the same as  immutable API that can't change in a patch release

Reply:
Taking another look, I thought that your suggested change would break the logic, but now I don't think it does.
So it does no harm to keep the method... I will try it

Reply:
This is no longer an issue, marking resolved